CoqIDE UI improvements and fixes

Frigory33 opened the pull request:
Hello. Thank you for providing such good libre software.

This week, I’ve worked on improving the UI of CoqIDE, especially the Preferences dialog. Indeed, I’ve started to use CoqIDE at the university, and two things shocked me: the lack of margins in the Preferences dialog, and the fact that by default you can’t use Ctrl-Right to move by word. So I decided to prepare a pull request, and thus I have spent hours and hours hacking CoqIDE and understanding lablgtk.

My changes (except very small ones) are described in my commit messages, I hope this is enough. Let me know if you have any questions.

Before submitting this pull request, I looked at the reported issues on this repository concerning CoqIDE. I was able to fix #16141 (I haven’t tested on Windows actually but I’m pretty sure I’ve fixed the bug) and #11024. I might have fixed #3977 too.

Usually, I’m doing many tests after having worked on each feature, so my changes quite rarely introduce bugs. For the most part, it’s really just UI.

A few notes:
- ~I have completely removed the Configwin modules. They were barely used; the preferences module did already implement most things by itself. So I went further in this direction. I added a Preferences_ui module. I moved the parts concerning UI from Preferences to Preferences_ui, and now the two modules have approximately the same size of 700 lines.~ This will be done in another PR, as comments below have asked for splitting this one.
- To suppress the problem of Ctrl-Right not working, I’ve set CoqIDE to use Alt by default (just like VsCoq). To me, it is much more appropriate. (I did not change the binding for macOS which is Ctrl-Cmd.)
- I’ve changed some other key bindings so that they are closer to standards and so that no key binding is used by more than 1 item. Undo is now bound to Ctrl-Z, Previous/Next tab to Ctrl-PgUp/PgDn (actually GTK offers Ctrl-Alt-PgUp/PgDn by default).

I hope you’ll like my changes.

jfehrle reviewed (commented):
Thanks for the PR.  CoqIDE needs more TLC (tender loving care) and some feature improvements.  Let us know if you're interested in making other improvements.  Also I'd be happy to address your questions about CoqIDE.

My initial thoughts:
- It's better to submit multiple unrelated changes as separate PRs for each to facilitate review/discussion.  You've changed thousands of lines--more work for the reviewer to sort out which changes are which purpose.  As it is, none of your various changes will be merged until all of them are reviewed.  Also it will be easier for you to update your commits if necessary if they're not stacked up.
- We generally don't update formatting such as indentation unless we're changing the logic because it makes it considerably harder to track the substantive changes in the git log.  Looks like you have many of these.
- Nontrivial user-visible changes should be described in a changelog entry.  Use the `make-changelog.sh` script to create a file under `doc/changelog`.  Changing default key mappings seems like something that should at least be mentioned.  And may be surprising to some users.
- The documentation should generally be updated to reflect notable user-visible changes.  (Did you read the CoqIDE documentation?  It describes how to customize most key bindings BTW.)

I didn't take a close look at anything yet.

jfehrle commented:
PS We can review the code before the changelog and doc changes are there (though sometimes they are helpful for understanding the changes).

Frigory33 commented:
> It's better to submit multiple unrelated changes as separate PRs for each

Yes, I’ve read it in the contributing guidelines. But I don’t know how to handle this correctly, as my changes usually always affect the same bunch of source files.

>  As it is, none of your various changes will be merged until all of them are reviewed.

I am not in a hurry. Take 3 months if you need to. :slightly_smiling_face:

> We generally don't update formatting such as indentation unless we're changing the logic because it makes it considerably harder to track the substantive changes in the git log. Looks like you have many of these.

Yes… I suggest you only look at the general diff, rather than each commit’s diff; the general diff is not much polluted by code formatting changes. For preferences_ui.ml, you could use these commands to see that you can skip one or two hundreds of lines:
```sh
git show origin/master:ide/coqide/configwin_ihm.ml | diff -W 120 -y /dev/stdin ide/coqide/preferences_ui.ml | less
git show origin/master:ide/coqide/preferences.ml | diff -W 120 -y /dev/stdin ide/coqide/preferences_ui.ml | less
```
But the hard work will obviously be to read this file, as I modified nearly everything in it.

>  (Did you read the CoqIDE documentation? It describes how to customize most key bindings BTW.)

Yes (I noticed the coqide.keys file), but I think the initial configuration of a piece of software matters a lot.

jfehrle commented:
> But I don’t know how to handle this correctly, as my changes usually always affect the same bunch of source files.

What matters is whether it's the same lines in the same file.  git readily handles nonoverlapping changes within a single file.  If 2 PRs change the same line, you would rebase the second one after the first has been committed.

> I am not in a hurry. Take 3 months if you need to. 🙂

I don't have that kind of time to spend.  Everyone on the project is a volunteer and has other demands on their time.

>> We generally don't update formatting such as indentation

> Yes… I suggest you only look at the general diff, rather than each commit’s diff; the general diff is not much polluted by code formatting changes. For preferences_ui.ml, you could use these commands to see that you can skip one or two hundreds of lines:

I suggest removing the formatting changes.  Even if you format in one commit and reverse in a second commit, it still confuses the git log.

Frigory33 commented:
>  Let us know if you're interested in making other improvements.

I would like to translate the interface, using [gettext/gettext-camomile](https://ocaml.org/p/gettext/). It’s a French piece of software and it is all only English… I would like help for a few things: integrating gettext in the repository (I don’t know much about dune, etc.), knowing where to put translation source files (template as POT, translations as PO), setting automatic compilation of source files from PO to MO, loading the MO files. If nobody here already knows how to do some of these things, I’ll manage to do them myself.

> I suggest removing the formatting changes. Even if you format in one commit and reverse in a second commit, it still confuses the git log.

Right, I’ll work on restructuring my commits.

>  Also I'd be happy to address your questions about CoqIDE.

I think it’s not specific to CoqIDE, but: what are the plans for exporting Coq source code to LaTeX or something else? At the university, we’re studying [sequent calculus](https://en.wikipedia.org/wiki/Sequent_calculus) and being able to get well-formatted proof trees directly from Coq code.

ppedrot commented:
@Frigory33 I don't want to lower your enthusiasm, but I think it's fair to say that the unofficial policy about CoqIDE is that it's mostly a legacy piece of code. There are various reasons for that, both technical and social. Nowadays the core-dev work on UIs is concentrated on vscoq. As a result we try not to touch CoqIDE too much, except for glaring bugs and similar annoyances.

In any case I'd definitely recommend splitting this PR, otherwise the probability it gets merged converges exponentially to zero.

jfehrle commented:
> I would like to translate the interface, using [gettext/gettext-camomile](https://ocaml.org/p/gettext/). It’s a French piece of software and it is all only English

This would be a huge project that would impact much of the system.  If you want to proceed, I suggest making a proposal and discussing this with the core team since it would be a pervasive change.  (Let's address this PR first.)  I don't know if they would think this is worth doing now (or ever).

Which parts do you want to translate?  CoqIDE menus, messages generated by Coq, the documentation?  I don't see how doing anything less than all of these would be valuable to users.  How about standard library coqdoc?  We would want an approach that all GUIs could potentially share.  Furthermore we'd need a strategy to regularly update the translations, especially for the documentation, which evolves continuously (and needs a lot of work).

> I think it’s not specific to CoqIDE, but: what are the plans for exporting Coq source code to LaTeX or something else? At the university, we’re studying [sequent calculus](https://en.wikipedia.org/wiki/Sequent_calculus) and being able to get well-formatted proof trees directly from Coq code.

Can you be more precise/specific?
- "Coq source code", meaning .v files?  Or the proof contexts?
- "to LaTeX or something else" - you could capture the output from `coqtop`, but I expect you have something more usable in mind
- "well-formatted proof trees directly from Coq code" - truly graphical trees or maybe something simple relying only on indentation?  [Prooftree](https://askra.de/software/prooftree/) created a graphical tree display for Proof General
- Do you want something totally automatic or that the user can control?  I think many like [Aelectryon](https://github.com/cpitclaudel/alectryon), but I don't know if it can be used as an automatic tool

Discussion of this should be outside of this PR (trying to follow multiple threads in a single PR or issue is unpleasant; people get confused and may give up on participating). Maybe open an issue to capture it.

jfehrle commented:
> I don't want to lower your enthusiasm, but I think it's fair to say that the unofficial policy about CoqIDE is that it's mostly a legacy piece of code. There are various reasons for that, both technical and social. Nowadays the core-dev work on UIs is concentrated on vscoq. As a result we try not to touch CoqIDE too much, except for glaring bugs and similar annoyances.

The main technical issue is that CoqIDE uses GDK V3 for the GUI, while V4 is the current version.  Converting to V4 is a major project that no one has chosen to take on.  Furthermore, as you may have discovered, there's no documentation of its OCaml interface, which differs in its details from the C interface and only supports part of GDK functionality.  Furthermore, GDK is buggy.

Another consideration is that CoqIDE is not based on any existing IDE.  In contrast `vsCoq` is based on `vsCode`, so it inherits features, ongoing improvements and bug fixes that we would have to implement ourselves in CoqIDE.  So not a great use of limited developer resources.  Nonetheless it's still a useful GUI.  I expect many incremental improvements will be accepted.  And it has certain benefits for prototyping new features.

jfehrle commented:
@Frigory33 How about starting by splitting off the following into a separate PR?  It seems like it's one of the simpler improvements.

- Document tabs are now reorderable - we'll need to check whether rearranging the tabs changes their index as used in `Wg_notebook.get_nth_term`.  I believe the debugger relies on the index not changing.  The debugger may not handle the closing a buffer case correctly.  Additional code changes may be needed.

Separately, "Best Fit" doesn't work sensibly for me.  The font is too small.  How do you define "Best Fit"?

![image](https://user-images.githubusercontent.com/1253341/228036428-cbf28249-e3da-41f8-ba35-43b9b403af4e.png)

Frigory33 commented:
> In any case I'd definitely recommend splitting this PR

You’re right, I’ll do it. Give me a few days.

> I think it's fair to say that the unofficial policy about CoqIDE is that it's mostly a legacy piece of code. There are various reasons for that, both technical and social. Nowadays the core-dev work on UIs is concentrated on vscoq.

I was presuming it was the case. But I don’t like VS Code, and CoqIDE is presented to us at the university, therefore I think it deserves a few improvements, at least to give a nice feeling to students. Furthermore, I like improving UX, I wanted to have some experience with GTK, and doing this with OCaml was very nice.

> This would be a huge project that would impact much of the system.

I don’t think so… If gettext is too heavy, I could just… Well, we’ll see this later, yeah.

>> what are the plans for exporting Coq source code to LaTeX or something else?
>
> Can you be more precise/specific?

I thought the *File > Export to* feature of CoqIDE was close to what I asked for, but when I tested it, it generated no proof trees. So I thought it was a work in progress. It’s not?

I mean generating proof trees just like the ones you can see on Wikipedia, e.g in [this section](https://en.wikipedia.org/wiki/Sequent_calculus#The_system_LK), no matter the input (thus it could use the `.v` code, `coqtop`’s output, rely on indentation; never mind, I just want it to generate nice proof trees), for an extract of code from `Lemma` to `Qed` or something like that.

Our teacher seems to write his proof trees in LaTeX. (1) I’m not used to LaTeX and (2) we don’t use CoqIDE so much and mostly have to write on paper sheets, but I don’t like paper, handwriting makes less beautiful trees and proving from CoqIDE is easier. I would like to see my proofs made in Coq in the handwritten format, to learn for the exam. Also, my mates at the university don’t care much about Coq code and I could share my homework with them more easily if I could convert it to our usual proof trees.

Anyway, I won’t have the tool I want before the exam comes, but I am thinking to future students. Maybe there is already a tool which does exactly what I want?

> Do you want something totally automatic or that the user can control?

A LaTeX or MathML source code would give me enough control.

> The main technical issue is that CoqIDE uses GDK V3 for the GUI, while V4 is the current version. Converting to V4 is a major project that no one has chosen to take on. Furthermore, as you may have discovered, there's no documentation of its OCaml interface, which differs in its details from the C interface and only supports part of GDK functionality. Furthermore, GDK is buggy.

From what I’ve seen, lablgtk3 supports very few of GDK (whose C documentation is not good either), but a fair amount of GTK. lablgtk3 also offers GtkSourceView 3, while the current version is 5. To me, GTK 3 isn’t that obsolete (even though upgrading is quite important to me)… But then, what should be used to make GUIs with OCaml? Actually, I used lablgtk’s documentation a lot; I already had some experience with GUI libraries, so for the most part, I just had to know how lablgtk’s classes are structured and what they offer.

I guess sometime I will make my own OCaml port of a GUI library (for now, wxWidgets would be my first choice).

> Separately, "Best Fit" doesn't work sensibly for me. The font is too small. How do you define "Best Fit"?

I didn’t make the algorithm, I just added one or two lines to make it have an effect. At first, I thought it was a misnomer for ‘Default zoom/Zoom 100%,’ but it is not, as Zoom in and Zoom out actually change the default zoom/font size. What I’ve understood is that it selects the font that makes all line fit in your editor’s width, so when you use this feature, the font size is determined by the length of the longest line in your buffer (I didn’t test it when word wrap is enabled though, I have to do it): the longer it is, the smaller the text will be. The line I added is [here (line 1242 in coqide.ml)](https://github.com/coq/coq/pull/17421/files#diff-55fae58ed643b31c58eab6b2f6ca928b5718eb0c288609c4efd1d29b8c203634R1242).

jfehrle commented:
I should have written GTK, not GDK :-)

"Export" appears to generate a nicely-printed version of the script in a file in the same directory as the named buffer using coqdoc.  If you check the git logs, it's old functionality.

Zimmi48 commented:
@Frigory33 It might be useful for you to join one of the upcoming [Coq Calls](https://github.com/coq/coq/wiki/Coq-Calls) to have a direct discussion with Coq developers about the future of CoqIDE. Would that time (Tuesday, 4pm, French time) work for you? If yes, just add an item to the next Coq Call page (create it if it doesn't exist yet) when you are available (with at least one-day notice).

jfehrle commented:
> join one of the upcoming [Coq Calls](https://github.com/coq/coq/wiki/Coq-Calls) to have a direct discussion with Coq developers about the future of CoqIDE

I'd be interested in participating.

Frigory33 commented:
> Would that time (Tuesday, 4pm, French time) work for you?

That’s a bit too soon for me. I’ve set the meeting to 5pm; I hope this is okay for other people.

> have a direct discussion with Coq developers about the future of CoqIDE

I haven’t planned to work on CoqIDE for a long time, but I’m okay for having a discussion. :slightly_smiling_face:

Frigory33 commented:
Well, I’ve just worked on reorganizing this PR. Each commit is much easier to analyze, and I haven’t put my main work (on the preferences dialog) in it.

> - Document tabs are now reorderable - we'll need to check whether rearranging the tabs changes their index as used in `Wg_notebook.get_nth_term`. I believe the debugger relies on the index not changing. The debugger may not handle the closing a buffer case correctly. Additional code changes may be needed.

I think the problem wasn’t (or not only) with the debugger, but you were right, I wasn’t careful with this and it broke a few things. This is now managed correctly, but with a quite complicated code which uses the widgets’ “oid” to recognize them (I didn’t find a better way of doing things).

Actually, what I wanted to do with the tab was to add a popup menu with a close menu item, and closing tabs with middle click. But I haven’t been able to do that.

ppedrot commented:
@Frigory33 you'll have to split it nonetheless, there are unrelated changes in this PR.

Nte that some are potentially problematic, namely the switch of default navigation modifiers to make Control displacements work as expected on the buffer. I remember this was a can of worms we opened a few years ago when it was decided which default modifiers should be taken. It turns out that every single desktop environment has its own convictions when it comes to the behaviour of arrows + some random combination of modifiers from {Ctrl, Shift, Alt, Meta}. Some use them for desktop switching, other for windows displacement, and more. We came up with the current status quo as a middle ground that would mostly work on all OSes, at the cost of losing the Control+arrow shortcut for buffer manipulation.

Frigory33 commented:
> It turns out that every single desktop environment has its own convictions when it comes to the behaviour of arrows + some random combination of modifiers from {Ctrl, Shift, Alt, Meta}. Some use them for desktop switching, other for windows displacement, and more.

I’ve never seen Alt-arrow used by any desktop environment. Alt-Left and Alt-Right were used by CoqIDE for tab switching, but this is very non-standard — Alt-Left and Alt-Right are usually mapped to “Back” and “Next” (in browsers and file managers); Alt-Up is mapped to “Parent directory”. Furthermore, I’ve just copied the choice made for VsCoq. I suppose Alt-arrow was initially used for tab switching, this is why it wasn’t used for navigation. If the desktop environment uses Alt-arrow, well, people will have to use the toolbar or remap the accelerators in this exceptional case — this is why such a setting must exist actually: for *exceptional* cases, not for the main one. The main case is: Ctrl-Right, Ctrl-Home and Ctrl-End are very common for navigation in a text editor. (In macOS, if I remember correctly, it’s Alt-arrow to move by word, but I haven’t changed the modifier for this OS: Ctrl-Cmd-arrow.)

Furthermore, there were other oddities: Ctrl-U for Undo (actually GtkSourceView automatically maps Ctrl-Z and Ctrl-Shift-Z, but I think it’s better to just show these accelerators instead of making a new one), and sometimes Ctrl instead of Command on macOS (as we can see [here](https://support.apple.com/en-us/HT201236), Ctrl is quite rarely used on macOS). There were also some overlapping cases. It was enough for me to think that the work on choosing the accelerators was not very rigorous, and thus my rigorous vision would be better.

> you'll have to split it nonetheless, there are unrelated changes in this PR.

Thank you for the remark, but can you be more precise? I don’t really see the point of splitting the PR; I feel it’s just more work for everyone… I can think of splitting like this:
- Not about programming:
  - [Add .editorconfig (useful for many editors)](https://github.com/coq/coq/pull/17421/commits/91a82d13e6a58982d0f7f56861c9200e2919a7d6)
  - [Update browser commands: use xdg-open for Linux, do not use “openUrl” and “-remote”](https://github.com/coq/coq/pull/17421/commits/088d17d2eb6048014d42a2c83d093127d56632c1)
  - [coqide FAQ: Update; format with Markdown](https://github.com/coq/coq/pull/17421/commits/584a762bd4c2a6a5d04c0c71246b7130b80d618e)
  - Changelog
- Miscellaneous GUI improvements and fixes:
  - [coqide: Improve find/replace UI (margins, icons)](https://github.com/coq/coq/pull/17421/commits/81b74eecf0461454855a57af968f8451f9ca93ba)
  - [coqide: A few fixes in file management](https://github.com/coq/coq/pull/17421/commits/669f11731e0a82f47c4e6d29d54229ef33f8b10f)
  - [coqide: Fix “Zoom fit” feature, which had no effect (makes all lines fit in editor’s width)](https://github.com/coq/coq/pull/17421/commits/b14a2836321223373986d7a76e6c3cd6ec5b2389)
  - [coqide: Fix document tabs position not being changed when the preference is applied](https://github.com/coq/coq/pull/17421/commits/f9466ce3e6325ff9e4b086a3e65ebbc4ab8afa0a)
  - [coqide document tabs: Factorize code; make tabs reorderable](https://github.com/coq/coq/pull/17421/commits/0daf8fe4c3797b3a14d7b190bbc2925df750f029)
  - [coqide: Move question_box from Configwin_ihm to a new Ui_dialogs module and allow buttons with stocks instead of plain labels](https://github.com/coq/coq/pull/17421/commits/39ab097b1d0a0ee99aaeb3fbbcef40b303d52813)
  - Changelog
- Menu bar, toolbar and accelerators:
  - [coqide: Improve menu items and accelerators](https://github.com/coq/coq/pull/17421/commits/119f842280b9ecb3a6249a482dea2125dc9cad4b)
  - [coqide: Add “Select All” item in the Edit menu](https://github.com/coq/coq/pull/17421/commits/f06748c828634f6ed86baf64bc22d902da18a304)
  - [coqide: Add icons to Debug menu items](https://github.com/coq/coq/pull/17421/commits/6725d00da42fdf671779cedf67bc660d03b6fa45)
  - [coqide: Put icon at a better place in Help menu](https://github.com/coq/coq/pull/17421/commits/3e93133026e1aba3166e0b4b1ca5ee3434caac9f)
  - [coqide: Apply View modifiers changes to the “Coq.PrintOpt” menu items only](https://github.com/coq/coq/pull/17421/commits/db285e0bf108c6bf0f883107b60ae8e98e07cb46)
  - Changelog

Would this be okay? :slightly_smiling_face:  
I’m quite disappointed to see how much work is needed to organize commits — when I’m working, I don’t have this rigor at all; my changes are not linear: I tend to work on many things simultaneously. I didn’t really see things like this — but as long as you have clear expectations and acknowledge the work, that’s cool. I hope it will be easier for me when I’ll have more experience.

jfehrle commented:
@Frigory33 See this Zulip conversation: https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs-.26-plugin-devs/topic/CoqIDE.20future

What is your Zulip ID?

BTW, I'd be happy to talk with you more if you wish.  Just let me know.  Coq Calls can be a bit chaotic.

jfehrle commented on ide/coqide/coq.ml:
```diff
@@ -672,7 +672,7 @@ struct
     { opts = [all_basic;existential;universes]; init = false;
       label = "Display all _low-level contents" };
-    { opts = [unfocused]; init = false; label = "Display _unfocused goals" };
+    { opts = [unfocused]; init = false; label = "Display un_focused goals" };
```
Tangent: the underline navigation mechanism is really weird.

![image](https://user-images.githubusercontent.com/1253341/230509046-b2c4f143-98d2-455b-9a6b-c20055e366e5.png)

The letter doesn't have to be unique.  Without the change, you'd just press `u` twice to get to the entry.  OTOH, I don't see how this form of menu navigation is all that useful:
- Getting the underlining to show in the menu is quirky.  AFAICT, you have to click on a top menu, then type a letter.  If the letter isn't assigned (e.g. `a`), they won't appear unless you navigate with a right/left arrow (among other quirks).
- You can't activate these directly from the editor panel or, e.g., the Messages panel (read-only).
- The functionality mostly duplicates hot key sequences (e.g. Shift+Alt+F)
- It's hard to pick out the underlined characters.

I wonder if there's a way to disable it--though I'd ask others' opinions first.

Frigory33 replied:
> The letter doesn't have to be unique.

You should open the current CoqIDE version and see what it does: this mnemonic letter is also used to determine the accelerator (default is: Shift-Alt-U, if U is chosen as a mnemonic) actually. See:
https://github.com/coq/coq/blob/584a762bd4c2a6a5d04c0c71246b7130b80d618e/ide/coqide/coqide.ml#L1271
Furthermore, for efficiency, it’s still better to choose distinct mnemonics in a menu item. I think it has importance as most GUI pieces of software still provide mnemonics. For instance, in Ms Office with the ribbon, you can still access the ribbon’s elements using Alt-letter, letter, letter, and they sometimes mention it in help contents if I remember well. Mnemonics also have historical importance: when GUIs appeared, while mouse devices were quite rare, mnemonics were the common way to quickly access menu items. Personally, I do use mnemonics for some items for which there isn’t a shortcut, notably in context menus: right-click, letter; or on buttons: Alt-letter; and my action is done in a fraction of a second without any effort. It is also helpful if the existing accelerator is borrowed by the desktop environment, for example.

To summarize: mnemonic letters are a very standard, easy-to-deal-with, non-bothering feature of GUIs. I hope this is enough to convince you :slightly_smiling_face:.

> It's hard to pick out the underlined characters.

At least, it has no cost to provide such mnemonics, no matter whether the choice is definitely the best or not, so let’s do it.

> Getting the underlining to show in the menu is quirky. / You can't activate these directly from the editor panel or, e.g., the Messages panel (read-only).

I don’t see what you mean; you may have behaviors which are specific to your desktop environment. Formerly (i.e. before Windows 95), mnemonics were shown all the time. Strangely, it seems to be the case in my Firefox and Thunderbird too. The standard way to use mnemonics is to press the underlined letter while holding Alt. Therefore, the underlining only appears when you press Alt. Historically, you could also press F10 to activate the menu bar, and Shift-F10 to show the context menu (this does not work in CoqIDE because these key shortcuts are borrowed by Debug menu items); now, there is the special Menu key for the context menu. Under Windows, you may also press Alt, letter instead of Alt-letter (meaning you can press Alt then release it, then press the letter, instead of pressing both at the same time), and F10 does not open the first menu (usually “File”) but only selects it.

When I was a child, my first feeling about these mnemonics was that they were not so useful, too. But then I discovered the option in the so-called “Folder options” on the Windows explorer: “Do not show rapid access keys before I press Alt” (I think it was disabled by default before Windows XP). Indeed, mnemonics are much more useful when you know they work with Alt :slightly_smiling_face:.

In the “Find in page” bar of Firefox, some time ago you had “_Previous” and “_Next” buttons, so you could press Alt-P to “Find previous.” Now that they’ve been replaced by buttons with icons, I don’t know if there is a shortcut for this and it’s annoying.

The common character to put before mnemonics (in other GUI libraries) is the ampersand `&`, but GTK requires an underscore instead, I don’t know exactly why.

jfehrle replied:
I've never used or wanted such mnemonics.  Only accelerators, which I use sparingly.  Too easy to pick the wrong operation.

> I don’t see what you mean; you may have behaviors which are specific to your desktop environment.

The documentation should say something about how to access mnemonics (e.g. to press Alt; I was opening a menu and then pressing left/right to see the mnemonics).  Also the convention for how CoqIDE assigns mnemonics and accelerators. (I hope there is an underlying consistency.  What we have seems haphazard.)  Contrary to what many believe, GUIs are not self-documenting.  The behavior of mnemonics is not simple to discover.

I think linking the accelerator to the mnemonic is problematic.  And it's too easy to bind an accelerator to multiple items (e.g. for CTRL); there should be a runtime error for that.  Nor do I see underscore for the File menu entries.

```
  menu file_menu [
    item "File" ~label:"_File";
    item "New" ~callback:File.newfile ~stock:`NEW;
    item "Open" ~callback:(File.load ~parent:w) ~stock:`OPEN;
```

But let's get back to the task at hand.

Frigory33 replied:
> I was opening a menu and then pressing left/right to see the mnemonics)

Okay. I thought I tried this and it didn’t work… But indeed it works. But it is not very useful if you have to use your mouse first.

> The documentation should say something about how to access mnemonics (e.g. to press Alt

We can mention it somewhere if you want. But please keep in mind that this is really a standard. By this, I mean: just as when you learn that Shift is to write capital letters and characters at the top of keys, that AltGr is for  or that Ctrl is for main hotkeys, or that there are navigation keys (too few people know about Home, End, PgUp, PgDn), you should learn that Alt is primarily for mnemonics. CoqIDE is not responsible here; the problem is that information on these standards hasn’t spread yet. People have to be more serious when teaching how to use a computer. When you have these keys so many hours in a week, it is important to know what they are useful for!

jfehrle commented on ide/coqide/coqide.ml:
```diff
@@ -1413,34 +1417,36 @@ let build_ui () =
   menu file_menu [
     item "File" ~label:"_File";
     item "New" ~callback:File.newfile ~stock:`NEW;
-    item "Open" ~callback:(File.load ~parent:w) ~stock:`OPEN;
+    item "Open" ~callback:(File.load ~parent:w) ~stock:`OPEN ~tooltip:"Open file";
     item "Save" ~callback:(File.save ~parent:w) ~stock:`SAVE ~tooltip:"Save current buffer";
-    item "Save as" ~label:"S_ave as" ~stock:`SAVE_AS ~callback:(File.saveas ~parent:w);
+    item "Save as" ~stock:`SAVE_AS ~callback:(File.saveas ~parent:w);
```
Underlined but no hot key.  As I say elsewhere, the underline mechanism seems broken and rather pointless.

![image](https://user-images.githubusercontent.com/1253341/230509712-4cbd009c-c99b-4578-8950-1352b8bdb540.png)

Frigory33 replied:
I think there is no real standard hotkey for this. Intuitively, I would have chosen Alt-S (because “Save as” is just like an alternative way of saving), and Shift-Ctrl-S for “Save all” (because saving all is like an elevated way of saving), but GTK apps apparently use Shift-Ctrl-S for “Save As.” So you would opt for Shift-Ctrl-S? I rarely use the “Save as” feature, I will rather close my editor program and move my file with my file manager (or the command line). If you need it, Alt-F, A can do the trick; to me, it’s enough.

jfehrle replied:
It doesn't have to have an accelerator.

jfehrle commented on ide/coqide/coqide.ml:
```diff
@@ -1514,10 +1517,10 @@ let build_ui () =
         radio "Set removed diff" 2 ~label:"Show diffs: added and _removed";
       ];
-    item "Show Proof Diffs" ~label:"_Show Proof (with diffs, if set)" ~accel:(modifier_for_display#get ^ "S")
+    item "Show Proof Diffs" ~label:"_Show Proof (with diffs, if set)" ~accel:"<Shift>F2"
```
Users tend to find unexplained changes to key bindings annoying.

Frigory33 replied:
The reason is: Shift-Alt-S is borrowed by “Hide synthesizable types” (not present in CoqIDE 8.16.1).

jfehrle replied:
Duplicates should generate an error message at startup.  Far too easy to create them.

Frigory33 replied:
You mean a developer would create duplicates too easily? GTK probably offers something to detect duplicates, but this would not be my way to work. When I add something to the interface, I tend to start exploring and playing again with everything to check whether it is well integrated and offers a new good global and slightly modified experience. This is relaxing and social. Therefore I’m not in favor of this addition you suggest; in my opinion, the project was not missing such a check, it was missing someone who is enough keen on UX.

jfehrle commented on ide/coqide/coqide_ui.ml:
```diff
@@ -98,9 +99,11 @@ let init () =
 \n    <menuitem action='Forward' />\
 \n    <menuitem action='Backward' />\
 \n    <menuitem action='Run to cursor' />\
+\n    <menuitem action='Reset' />\
```
Did you intend to move Reset?  If so, why?

Frigory33 replied:
Yes I did. One goal of a good GUI (as opposed to a CLI) is to make the user able to quickly understand and master many commands or features. When I saw “Reset” at this place, despite the icons and the corresponding hotkeys, it took me some time to make sure that it was really just the opposite of “Run to end,” because the labels really differ and there were other items in between (and even more: “Reset” was at the end of the “Navigation” items, so it could stand for something completely different). Moving it near “Run to end” solves this problem.

Another reason to this move is how I consider what you should do when discovering CoqIDE: you will start by moving forward or run to cursor (if your teacher tells you to skip what’s at the top), maybe you will go backward to focus on something (how it is before, how it is after, let’s see it again and again), and then you should reset everything (provided that you are not asked to skip what is at the top, but in the course I followed, there was rarely something to skip) to make sure you have understood everything well. Other commands will find their meaning later (in my course, we had no reason to use “Fully check” and “Interrupt”), so should be the last in the menu.

Putting “Reset” before “Run to end” might seem inconsistent with “Forward” being before “Backward,” but I think it’s better like that because “Run to end” might have something to do with “Fully check” and “Interrupt,” whereas “Reset” does not.

jfehrle commented on ide/coqide/coqide_ui.ml:
```diff
@@ -98,9 +99,11 @@ let init () =
 \n    <menuitem action='Run to cursor' />\
+\n    <menuitem action='Reset' />\
 \n    <menuitem action='Run to end' />\
+\n    <menuitem action='Force' />\
```
This doesn't show on the menu,  What is it supposed to do?

Frigory33 replied:
It is “Fully check” (the ID of this menu item is “Force”) and I don’t know what it is supposed to do actually. As it is present in the toolbar, I wanted to have it as a menu item, at least to show the accelerator.

jfehrle replied:
Then it should be "Fully Check" in the menu.  It is briefly described in the [this section](https://coq.inria.fr/distrib/current/refman/practical-tools/coqide.html#asynchronous-mode) of the documentation.

Frigory33 replied:
“Force” was just an identifier, this is why I didn’t bother changing it. The label was defined in coqide.ml. But OK, this is fixed: the ID is now “Fully check”.

jfehrle commented on ide/coqide/coqide_ui.ml:
```diff
@@ -98,9 +99,11 @@ let init () =
 \n    <menuitem action='Interrupt' />\
-\n    <menuitem action='Reset' />\
+\n    <separator/>\
 \n    <menuitem action='Previous' />\
 \n    <menuitem action='Next' />\
```
FWIW If you read the code, this searches for the next/prev instance of the current word in the buffer.  This should be integrated into the general text search; use CTRL F for next and SHIFT CTRL F for previous.

Frigory33 replied:
Yes, I’ve seen this feature described somewhere (in a bug report, maybe). So I added separators… I don’t know in which case this feature would be so useful; as it was put in the toolbar, I tend to think that someone finds it very, very useful. But maybe the procedure was just to put all Navigation menu items in the toolbar.

Shift-Ctrl-F is usually used for “Find in Files” (or “Find in all opened tabs”). Ctrl-F is used by the common “Find / Replace.” So I don’t understand your idea. Do you want these menu items to be put in the Edit menu (under the “Find Previous” menu item)? I think it’s correct to have them in the Navigation menu.

jfehrle replied:
I didn't explain in full.  I think searching should work more like #14040.  If the search bar isn't shown, Ctrl-F shows it.  If the search bar is shown, Ctrl-F means to go to the next item.  It could also be Ctrl-F3 and F3.  I don't know when or if I will get to implement that, though.

> Shift-Ctrl-F is usually used for “Find in Files” (or “Find in all opened tabs”).

A very subjective statement.

Frigory33 replied:
> If the search bar is shown, Ctrl-F means to go to the next item. It could also be Ctrl-F3 and F3.

Currently, it is F3 and Shift-F3, which is standard (it works in Firefox for instance) and works whether or not the dialog is shown. So this is in place.

> A very subjective statement.

This is also my manner to say that I agree with this convention I’ve seen used in at least a couple of apps.

> I don't know when or if I will get to implement that, though.

I can do it if you want, though I want to say that the current dialog is more my style, because I like the very explicit text labels. I don’t think your idea will improve the productivity of people. But if it’s important for some important CoqIDE users, it’s still useful.

jfehrle commented on ide/coqide/coqide_ui.ml:
```diff
@@ -160,15 +164,18 @@ let init () =
 \n  <toolitem action='Forward' />\
 \n  <toolitem action='Backward' />\
 \n  <toolitem action='Run to cursor' />\
+\n  <toolitem action='Reset' />\
```
Reset should remain where it was.  It's an infrequent operation.  `Force` should go away,

Frigory33 replied:
I’m not sure it’s so infrequent. To me, it has an important (and common) role: *clear everything*, i.e. clear the panels, remove the highlight, rest the mind, and overcome bugs in CoqIDE (in my class, some of us had to Reset the script execution because CoqIDE stuck at some point). I think it is psychologically important to quickly know how to reset everything. It helps people trust the program and have hindsight.

But more important is: the order of items must not be determined by habits of experienced users, but by how much it will be easy to understand by new users. I think the order of items doesn’t matter for experienced users: they should rather use hotkeys. If you’re still using buttons and menu items when you’re experienced (except for demonstrating something to somebody), I’m sad for you…! Even if you are, the order doesn’t matter because you know where they are. These GUI components are mostly meant for people who don’t know the commands by heart. Striking a correct balance between (1) ensuring to advanced users that they will always be an efficient way to use the program, and (2) making the program easy to understand by (and, in some way, playful for) new users, is something very important to me. I feel program designers always focus either on (1) or (2), the other point being really forgotten. It tires me and even makes me angry, because it enforces gaps between experienced people and laypeople, which is very bad for society.

I could think about another organization: moving the Interrupt closer to the beginning of the list, e.g.:
Forward | Backward | Run to cursor | Interrupt | Fully check | Run to end | Reset
But I think it is actually good to have Interrupt at the end because it’s something that you will use in a quite negative case (i.e. when computing is too long or turns out in an infinite loop — I don’t know if we can have infinite loops in Coq). So it’s psychologically a good choice to have it at the end; it helps have the mind well-structured.

(About “Force,” see previous reply.)

jfehrle replied:
We'll figure something out, but I think there are higher priority things to fix than menus, mnemonics and accelerators.  Some features are unusable, others just don't make much sense.

Frigory33 replied:
It was quick to fix and to me this is an important base to have before working on more complex things. This is not complex because it is very classical; complexity will be in things which are specific to our piece of software.

I’m not aware of these unusable or inappropriate features…! I will be interested in improving and fixing.

jfehrle commented on ide/coqide/coqide.ml:
```diff
@@ -1413,34 +1417,36 @@ let build_ui () =
-    item "Redo" ~stock:`REDO
+    item "Redo" ~accel:"<Primary><Shift>z" ~stock:`REDO
       ~callback:(cb_on_current_term (fun t -> t.script#redo ()));
+    item "Select All" ~accel:"<Primary>a" ~stock:`SELECT_ALL
+      ~callback:(cb_on_current_term (fun t -> t.script#select_all ()));
```
This breaks selecting the contents of the Goal or Messages panel with CTRL A.

Frigory33 replied:
Thank you. I think I was able to make it handle all the cases.

jfehrle commented on ide/coqide/coqide.ml:
```diff
@@ -1598,12 +1601,16 @@ let build_ui () =
-    item "Continue" ~label:"_Continue" ~accel:"F9" ~callback:Nav.continue;
-    item "Step in" ~label:"Step in" ~accel:"F10" ~callback:Nav.step_in;
-    item "Step out" ~label:"Step out" ~accel:"<Shift>F10" ~callback:Nav.step_out;
+      ~stock:`MEDIA_RECORD ~callback:MiscMenu.toggle_breakpoint;
+    item "Continue" ~label:"_Continue" ~accel:"F9"
+      ~stock:`MEDIA_NEXT ~callback:Nav.continue;
+    item "Step in" ~label:"Step in" ~accel:"F10"
+      ~stock:`MEDIA_PLAY ~callback:Nav.step_in;
+    item "Step out" ~label:"Step out" ~accel:"<Shift>F10"
+      ~stock:`MEDIA_FORWARD ~callback:Nav.step_out;
```
I don't see these icons on my system.  Are you sure they're available on all platforms?

![image](https://user-images.githubusercontent.com/1253341/230512308-4f3af8d3-052c-47b0-aea4-3641ff4e4a08.png)

Frigory33 replied:
Here are the icons I have:
![coqide-icons-debug-help](https://user-images.githubusercontent.com/5684776/232521885-da79c551-bb5f-450f-8e44-da58e789b7c7.png)

> Are you sure they're available on all platforms?

I’m not a specialist of GTK, I don’t really like it (as I said above, wxWidgets follows my ideas and principles much better). The icons you have depend on your GTK icon theme; I personally use the GNOME icon theme (I never use GNOME though). I can’t provide much more support to people who don’t have the icons…! I’m using Manjaro Linux with Xfce and I can choose my icon theme in Xfce Settings Manager > Appearance > Icons. The first icon is stored on my system at: /usr/share/icons/gnome/24x24/actions/media-record.png

jfehrle replied:
I'm not sure the icons add much.  These don't convey the meaning of the operations well.  For example, your icon for "step in" is more appropriate for "continue".  Ideally we would make some of our own icons.  Or just omit them.

Frigory33 replied:
I think it’s still better to have these icons when they work; it adds some colors, some flavor. Without them, the feature may look cold, distant, or even unmaintained.

> your icon for "step in" is more appropriate for "continue".

I don’t think so, because “Continue” may go immediately to the end of the buffer, while “Step in” will advance only of a minimal step.

> Ideally we would make some of our own icons.

Indeed, if it is a matter of a few shapes, I’m able to produce the SVG and PNG image files.

jfehrle commented on ide/coqide/coqide.ml:
```diff
@@ -1614,13 +1621,13 @@ let build_ui () =
   menu help_menu [
     item "Help" ~label:"_Help";
-    item "Browse Coq Manual" ~label:"Browse Coq _Manual"
+    item "Browse Coq Manual" ~label:"Browse Coq _Manual" ~accel:"<Shift>F1" ~stock:`HELP
```
I don't see any icons on the Help menu on my system (Windows/WSL with Ubuntu/XWindows).

jfehrle commented on ide/coqide/preferences.ml:
```diff
@@ -310,45 +310,48 @@ let automatic_tactics =
-  pref#connect#changed ~callback:cb
+  ignore (pref#connect#changed ~callback:cb)
+
+let select_arch m m_osx =
+  if Coq_config.arch = "Darwin" then m_osx else m
```
Seems too mac-centric.  Maybe should be a bit more general.

OTOH, I tend to agree with Pierre-Marie that remapping keys is a losing game, perhaps better left to each user at this point.

Frigory33 replied:
> Seems too mac-centric. Maybe should be a bit more general.

This is not my code, I just moved the lines for this to be less messy. On macOS, things are very different:
- “Command” (Super) seems to have the role of “Ctrl” (for common hotkeys);
- “Option” (Alt) seems to have the role of “Ctrl” (for moving by word) or “AltGr”;
- “Ctrl” seems to have the role of “Super”.

Thus I think distinguishing between macOS and other systems like this is fully justified. GTK offers “Primary” as a way to designate macOS's Command and other systems’ Ctrl at the same time (even web browsers don’t do that); but sometimes, this is not enough.

> OTOH, I tend to agree with Pierre-Marie that remapping keys is a losing game, perhaps better left to each user at this point.

Many users don’t bother browsing offered settings. The initial configuration is very important. But yeah, developing something to ease changing hotkeys could be a good idea.

jfehrle replied:
Unfortunately, if a key is not bound in `coqide.keys`, it gets the default setting.  If we change the default, then any user who had the default suddenly has the new value.  Not so friendly.

Frigory33 replied:
Sure, I haven’t planned to do it often. In this case, I think people will be able to understand once that default key bindings were a bit messy.

jfehrle reviewed (commented):
Some initial comments on the first 8 commits as I try to understand what you've actually changed.

jfehrle reviewed (commented):
Comments on the remaining commits.  In general it looks good, a few things for you to attend to.

I want to look at the reorderable tabs commit more closely.
I'll respond to your ideas for grouping the commits into PRs a bit later.
Also, let me think a little more about remapping key assignments.

Feel free to push back on my comments.  Some of them are merely observations.

What did you think of the Coq Call?  We tend to get into lengthy discussions/arguments, perhaps not so fun to participate in.  We didn't get into some of the questions that you're eager to address, i.e. internationalization and generating proof trees.  I hope you want to make other improvements after we wrap up this PR.

Frigory33 commented:
Hello @jfehrle, thank you for your comments and replies. I’ve been busy for a few days and it will continue until next Friday (I’m working on a project for university, and the deadline is on Friday). I’ll be happy to talk with you after this time.

Frigory33 commented on ide/coqide/configwin_ihm.ml:
```diff
@@ -679,8 +679,8 @@ let edit ?(with_apply=true)
       match dialog#run () with
-        | `APPLY  -> config_box#apply; iter Return_apply
-        | `OK -> config_box#apply; destroy (); Return_ok
+        | `APPLY  -> config_box#apply; apply (); iter Return_apply
+        | `OK -> config_box#apply; apply (); destroy (); Return_ok
```
No, see:
https://github.com/coq/coq/blob/584a762bd4c2a6a5d04c0c71246b7130b80d618e/ide/coqide/coqide.ml#L1468

Frigory33 commented on ide/coqide/fileOps.ml:
```diff
@@ -72,11 +72,11 @@ object(self)
-                      "Disable Auto Revert"]
+            ~buttons:[ButtonWithLabel "Revert from file";
+                      ButtonWithLabel "Overwrite file";
+                      ButtonWithLabel "Disable Auto Revert"]
```
“Revert” means “forget the source code that is currently in the editor, and put the content of your Coq file (from which you got your Coq source or in which you last saved your Coq source) in the editor instead.” I’ve seen the feature evoked in the documentation along with the “External editor” feature: it is advised to enable auto revert if you use an external editor. I don’t really understand why they ask for a delay; when I tried to use it, CoqIDE seemed to ask me immediately when I activated the CoqIDE window if I wanted to revert, and I think most IDEs work like this. It is also common to have a “Revert” item in the File menu.

jfehrle replied:
A very poor choice of name.  The dictionary definition of "revert" is "return to (a previous state, practice, topic, etc.)".  "reload" is somewhat better.

Re-reading the file should be triggered by the CoqIDE window getting focus, but I'm not sure XWindows and/or GTK can do that.  (And writing should be triggered by CoqIDE losing focus.)

Frigory33 replied:
> "reload" is somewhat better.

I completely agree. In fact, it also exists as a menu item:
https://garrigue.github.io/lablgtk/lablgtk3/lablgtk3/GtkStock/index.html#type-gtk_stock_id.REVERT_TO_SAVED
where “revert” generally corresponds to reloading a file which has not been changed since the last loading. In French, we also had the wrong word “Rétablir” (revert) instead of “Recharger” (reload).

I’ve just batched-replaced all occurrences of “revert” to “reload”.

Frigory33 replied:
> Re-reading the file should be triggered by the CoqIDE window getting focus, but I'm not sure XWindows and/or GTK can do that.

It is a very common feature, so GTK or GDK probably have it, but lablgtk may not. We’ll see this later.

> (And writing should be triggered by CoqIDE losing focus.)

I don’t like this kind of automatic non-obvious behaviors. To me, the good practice is to get used to pressing Ctrl-S all the time :laughing:. I think such automatic behaviors lead people not to be aware of what their computer does, and (as said by many free software enthusiasts, but it may be even more important to me than to them) it is very bad. I also have a few practices which are not compatible with this feature, e.g. I will save something I want to test elsewhere, but before testing, I will change my buffer to what I want my code to be just after the testing (more concretely, this means I press: Ctrl-Z several times, Ctrl-S, Ctrl-Shift-Z until it has no effect, Alt-Tab, testing, Alt-Tab, Ctrl-S).

But obviously, I am not against providing this feature as a by-default-disabled preference.

jfehrle replied:
"obvious" is very subjective

Frigory33 replied:
In fact, not so much. A computer behavior is obvious when it is a technical implication, a technical optimization, a way to avoid some side-effect or dependency, or at least quite close to some very well-known standard.
- For instance, saving a source file before compiling it is standard, technically simpler, avoids depending on another location on which the editor has the permission to write, and is also quite an optimization because when you spend time compiling files and testing the result, you usually do want the corresponding source files being saved. *But* I’ve already seen a software program (I don’t remember which) which does save source files in a temporary location (/tmp) for compilation, if they are not already saved (and then saving your source file requires a bit more computer work).
- On the other hand, saving when losing focus is not at all technically implied, may cause useless computer work, may have the side-effect of overwriting something that the user wanted to keep… And we could even take in account further implications of automatic saving, e.g. if you save on a distant location and have a limited Internet connection.

I won’t blame it if you like this kind of auto-saving feature (which usually has no wrong implication at all), but at first it didn’t exist and I think there are good reasons to avoid the magic as a default behavior.

jfehrle replied:
The IDE I've been using for the last 20 years (IntelliJ IDEA) has always autosaved when it loses focus and reloads when it gains focus.  It also has great diff capabilities, so I typically review my changes compared to git (in the IDE) before I commit.  I can edit the changes or revert individual changes in the diff display.  (Reviewing your own code is very helpful!)  Undesired changes are usually very easy to pick out.  I can also revert all changes to a set of files.

Explicitly saving means you need to think about the editor state and the file system state separately.  For editing, I've not found that a useful distinction and a drawback if you ever forget to save.  But it's a matter of personal preference.

Frigory33 commented:
> What is your Zulip ID?

610722. I will post a reply to the conversation tomorrow (Tuesday).

I’m not available before 6pm (Paris time) tomorrow, so I would prefer to have the Coq call on Wednesday, is it possible? (I don’t know how to ask for it on Zulip.)
I will be available next week, but after (till the end of June), it may be harder because I will be doing an internship with one of my teachers (and it is not about proofs).

> Coq Calls can be a bit chaotic.

Yes, this is always my feeling in such meetings. It is even harder in English as I am very bat at listening comprehension. Even in French, it is hard: I will end being very tired if I try to follow everything — it might be because people tend to focus too much on their own needs and wishes, instead of deeply helping build a good consensus for the organization. But meeting people sometimes is still good.

> BTW, I'd be happy to talk with you more if you wish. Just let me know.

Written conversations are much more comfortable for me! I will think about what kind of conversation I would be eager to have. (I’m currently quite tired because of the end of the year at the university.)

Frigory33 commented on .editorconfig:
```diff
@@ -0,0 +1,10 @@
+root = true
```
Indeed. I hadn’t understood the scope of the changelog. Fixed.

Frigory33 commented on ide/coqide/ui_dialogs.ml:
```diff
@@ -0,0 +1,46 @@
+  | _ -> ()
+  in
+  iter_buttons 1 buttons;
+  let _ = window#connect#destroy ~callback:GMain.Main.quit in
```
Indeed, I messed with the possible syntaxes. It is now reverted.

Actually my preferred syntax for this is starting the line with `ignore @@`. It is used sometimes, mainly in coqide.ml. But `let _ =` is used much more.

Frigory33 commented on ide/coqide/fileOps.ml:
```diff
@@ -106,8 +106,8 @@ object(self)
   else
-    let answ = Configwin_ihm.question_box ~title:"File exists on disk"
-      ~buttons:["Overwrite"; "Cancel";]
+    let answ = Ui_dialogs.question_box ~title:"File exists on disk"
+      ~buttons:[ButtonWithLabel "Overwrite"; ButtonWithStock `CANCEL]
```
Actually, it looks like we can’t decide the content of this confirmation dialog which descends from the file dialog! This is a bit weird. Requesting for the dialog to appear is needed, but we have to provide a fake content which won’t be used. I’ve just added a comment in the code to make it clearer.

Frigory33 commented on doc/changelog/10-coqide/17421-coqide.rst:
```diff
@@ -0,0 +1,50 @@
+- **Added:**
+  .editorconfig file (used by many source code editors
+  to set things like indent size)
```
OK. I just looked at what `make-changelog.sh` did, then I copy-pasted the result many times…

I’m not sure about how it should be, I have changed it. Is it correct now?

jfehrle replied:
It looks better, but usually these files are shorter.  As I've said, this PR needs to be split, in part so others can take a look.  You might start with the first commit.

Frigory33 commented on ide/coqide/coqide.ml:
```diff
@@ -117,9 +117,12 @@ let make_coqtop_args fname =
+          | Some _ -> args
+          | None -> "-topfile"::fname::args
+        with Invalid_argument _ ->
+          failwith "CoqIDE cannot open files which don't have an extension in their filename."
```
Indeed. Thanks to another comment of yours, I’ve just noticed that CoqIDE can save to filenames which don’t have an extension, but cannot reopen them afterwards, so there is an amusing inconsistency.

Well, for now, I just wanted the message to be understandable by an expert without any effort. Solving other problems will require more work.

Frigory33 commented on ide/coqide/preferences.ml:
```diff
@@ -1060,25 +1063,19 @@ let configure ?(apply=(fun () -> ())) parent =
-      "firefox -remote \"openURL(%s,new-windows)\" || firefox %s &";
-      "seamonkey -remote \"openURL(%s)\" || seamonkey %s &"
-    ] in
+    let predefined = [Coq_config.browser; {|firefox "%s" &|}; {|seamonkey "%s" &|}; {|chromium "%s" &|}] in
```
The two parts of your sentence look contradictory to me… Netscape Navigator was stopped in 2008. These strings are just suggested choices; I didn’t even know that “mozilla” ever was a command, but it and “netscape” are surely now too old/rare to have their place here. It would probably be more sensible to add “safari” or “chrome”; but the default command will usually work. If people had already selected the “netscape” or “mozilla” command, it is written in their preferences file and will be reused even with this change.

I don’t think it is easier to read on separate lines… I like using the full length of lines, to see more code on my screen at once. I also have a tendency not to like other people’s coding styles sometimes, but I think we have to learn that a coding style is okay as long as it is coherent (meaning it follows precisely defined and determinable rules) — and we don’t have such a coherence in the CoqIDE source code.

Frigory33 commented on ide/coqide/FAQ.md:
```diff
@@ -0,0 +1,85 @@
+# CoqIDE FAQ #
+
```
I see it on my system at `/usr/share/doc/coqide/FAQ`. It looks quite outdated, but I think it’s still better with Markdown formatting.
